File: mailblastr-ruby/lib/mailblastr/audiences.rb
# frozen_string_literal: true

module Mailblastr
  module Audiences
    class << self
      # POST /audiences — params: { name: "..." }
      def create(params)
        Client.request(:post, "/audiences", body: params)
      end

      # GET /audiences/:id
      def get(audience_id)
        Client.request(:get, "/audiences/#{Client.path_escape(audience_id)}")
      end

      # GET /audiences
      def list(params = {})
        Client.request(:get, "/audiences", query: Client.pagination(params))
      end

      # Rename an audience. PATCH /audiences/:id
      def update(audience_id, params)
        Client.request(:patch, "/audiences/#{Client.path_escape(audience_id)}", body: params)
      end

      # DELETE /audiences/:id
      def delete(audience_id)
        Client.request(:delete, "/audiences/#{Client.path_escape(audience_id)}")
      end

      # Import contacts from a link-shared Google Sheet; header columns become
      # contact properties and rows land in a fresh segment.
      # POST /audiences/:id/contacts/import-sheet — params: { url:, segment_name: }
      def import_sheet(audience_id, params)
        Client.request(:post, "/audiences/#{Client.path_escape(audience_id)}/contacts/import-sheet", body: params)
      end
    end
  end
end

File: mailblastr-php/src/Resources/Audiences.php
<?php

declare(strict_types=1);

namespace Mailblastr\Resources;

use Mailblastr\Client;

/**
 * Audiences. Note the domain-first model: each sending domain also has its own
 * auto-managed contact POOL audience (its `domain` field names the domain).
 */
class Audiences extends Resource
{
    /**
     * Create an audience. POST /audiences
     *
     * @param array $payload ['name' => …]
     */
    public function create(array $payload): array
    {
        return $this->client->request('POST', '/audiences', $payload);
    }

    /** Retrieve an audience. GET /audiences/:id */
    public function get(string $id): array
    {
        return $this->client->request('GET', '/audiences/' . Client::e($id));
    }

    /**
     * List audiences. GET /audiences
     *
     * @param array $params Cursor pagination: limit, after, before.
     */
    public function list(array $params = []): array
    {
        return $this->client->request('GET', '/audiences' . $this->paginationQuery($params));
    }

    /**
     * Import contacts from a link-shared Google Sheet. Header columns become
     * contact properties (usable as {{merge_tags}}); rows land in a fresh
     * segment. POST /audiences/:id/contacts/import-sheet
     *
     * @param array $payload ['url' => sheet URL, 'segment_name' => …]
     */
    public function importSheet(string $audienceId, array $payload): array
    {
        return $this->client->request(
            'POST',
            '/audiences/' . Client::e($audienceId) . '/contacts/import-sheet',
            $payload
        );
    }

    /**
     * Rename an audience. PATCH /audiences/:id
     *
     * @param array $payload ['name' => …]
     */
    public function update(string $id, array $payload): array
    {
        return $this->client->request('PATCH', '/audiences/' . Client::e($id), $payload);
    }

    /** Delete an audience. DELETE /audiences/:id */
    public function remove(string $id): array
    {
        return $this->client->request('DELETE', '/audiences/' . Client::e($id));
    }
}
